### ISSUE TITLE
CI test failed with self-signed.badssl.com/104.154.89.105

### ISSUE THREAD

**Issue by drwetter** [drwetter/testssl.sh#431]
Hi @dcooper16 

CI test from @seccubus failed, see below. I disabled the test for now. Curious why it said chain incomplete instead of self signed

This was triggered by the recent binary I uploaded but I suspect the check for self signed is not correct.

Thx, Dirk

```
t/01_badssl.com.t .. 

ok 1 - Running testssl.sh against badssl.com to craete a baseline (may take 2~3 minutes)
ok 2 - We have more then 10 findings
ok 3 - Running testssl against expired.badssl.com
ok 4 - The certificate should be expired
ok 5 - Finding reads expired.
ok 6 - Severity should be NOT ok
ok 7 - We had a finding for this in the JSON output
ok 8 - Running testssl against self-signed.badssl.com
ok 9 - The certificate should not be expired
ok 10 - Finding doesn't read expired.
ok 11 - Severity should be ok
ok 12 - We had a finding for this in the JSON output
not ok 13 - Chain of trust should fail because of self signed

#   Failed test 'Chain of trust should fail because of self signed'
#   at t/01_badssl.com.t line 53.
#                   '
# ###########################################################
#     testssl.sh       2.8rc1 from https://testssl.sh/dev/
#     (b1595c1 2016-07-26 21:03:09 -- 1.531)
# 
#       This program is free software. Distribution and
#              modification under GPLv2 permitted.
#       USAGE w/o ANY WARRANTY. USE IT AT YOUR OWN RISK!
# 
#        Please file bugs @ https://testssl.sh/bugs/
# 
# ###########################################################
# 
#  Using "OpenSSL 1.0.2-chacha (1.0.2i-dev)" [~183 ciphers]
#  on testing-worker-linux-docker-8f043e09-3467-linux-2:./bin/openssl.Linux.x86_64
#  (built: "Jun 22 19:32:29 2016", platform: "linux-x86_64")
# 
# 
#  Start 2016-07-26 19:06:23    -->> 104.154.89.105:443 (self-signed.badssl.com) <<--
# 
#  rDNS (104.154.89.105):  105.89.154.104.bc.googleusercontent.com.
#  Service detected:       HTTP
# 
# 
#  Testing server defaults (Server Hello) 
# 
#  TLS extensions (standard)    "server name/#0" "renegotiation info/#65281" "EC point formats/#11" "session ticket/#35" "heartbeat/#15" "next protocol/#13172"
#  Session Tickets RFC 5077     300 seconds (PFS requires session ticket keys to be rotated <= daily)
#  SSL Session ID support       yes
#  TLS clock skew               random values, no fingerprinting possible 
#  Signature Algorithm          SHA256 with RSA
#  Server key size              RSA 2048 bits
#  Fingerprint / Serial         SHA1 079B3259D07C4DE2A1CE0EF4A5B5599D3B2D62EA / 9CF4D567E8D3601C
#                               SHA256 745D3A6BB0CBC86AEC5AB97E0C5C5C9681F91B3C1BE0E7B06835B8F180A5CAD1
#  Common Name (CN)             "*.badssl.com" (CN in response to request w/o SNI: "badssl-fallback-unknown-subdomain-or-no-sni")
#  subjectAltName (SAN)         "*.badssl.com" "badssl.com" 
#  Issuer                       self-signed (NOT ok)
#  Trust (hostname)             Ok via SAN wildcard and CN wildcard (SNI mandatory)
#  Chain of trust               NOT ok (chain incomplete)
#  EV cert (experimental)       no 
#  Certificate Expiration       396 >= 60 days (2015-08-28 01:20 --> 2017-08-27 01:20 +0000)
#  # of certificates provided   1
#  Certificate Revocation List  --
#  OCSP URI                     --
#  OCSP stapling                --
# 
# 
#  Done 2016-07-26 19:06:27    -->> 104.154.89.105:443 (self-signed.badssl.com) <<--
# 
# 
# '
```

**Comment by drwetter** [comment]
Sorry, @dcooper16 . Culprit is in ``determine_trust()``, so not your recent commit.

Change of behaviour/return code  is strange though.

**Comment by dcooper16** [comment]
I can't explain the change of behavior, as I get the same results no matter what version of OpenSSL I use, but I traced the code in OpenSSL and found out the reason for the response.

The relevant code is the following in crypto/x509v3/v3_purp.c:
```
    if (!X509_NAME_cmp(X509_get_subject_name(x), X509_get_issuer_name(x))) {
        x->ex_flags |= EXFLAG_SI;
        /* If SKID matches AKID also indicate self signed */
        if (X509_check_akid(x, x->akid) == X509_V_OK &&
            !ku_reject(x, KU_KEY_CERT_SIGN))
            x->ex_flags |= EXFLAG_SS;
```
In the case of self-signed.badssl.com, the issuer and subject names match and the certificate does not contain any key identifiers (so `X509_check_akid` succeeds), but the certificate contains a keyUsage extension with the following value:
```
            X509v3 Key Usage: 
                Digital Signature, Non Repudiation, Key Encipherment
```
Since the keyUsage extension is present and does not assert the keyCertSign bit, `ku_reject` succeeds and the certificate is not marked as being self-signed.

**Comment by seccubus** [comment]
Maybe they changed the certificate?

**Comment by typingArtist** [comment]
Here, everything works. Perhaps this is an environment setup problem?

```
$ t/01_badssl.com.t 
ok 1 - Running testssl.sh against badssl.com to craete a baseline (may take 2~3 minutes)
ok 2 - We have more then 10 findings
ok 3 - Running testssl against expired.badssl.com
ok 4 - The certificate should be expired
ok 5 - Finding reads expired.
ok 6 - Severity should be NOT ok
ok 7 - We had a finding for this in the JSON output
ok 8 - Running testssl against self-signed.badssl.com
ok 9 - The certificate should not be expired
ok 10 - Finding doesn't read expired.
ok 11 - Severity should be ok
ok 12 - We had a finding for this in the JSON output
ok 13 - We had a finding for this in the JSON output
ok 14 - Chain of trust should be ok
ok 15 - Finding says certificate can be trusted.
ok 16 - Severity should be OK
ok 17 - We had a finding for this in the JSON output
ok 18 - Running testssl against incomplete-chain.badssl.com
ok 19 - Chain of trust should fail because of incomplete
ok 20 - Finding says certificate cannot be trusted.
ok 21 - Severity should be NOT ok
ok 22 - We had a finding for this in the JSON output
1..22
```

**Comment by drwetter** [comment]
@typingArtist : that's because I commented out test no 13

**Comment by drwetter** [comment]
@dcooper16 : thx for tracking this down. So basically openssl has a problem here and @badssl.com's test is thus not working.

Question @seccubus: You have a better idea for this particular CI test? Otherwise I could setup s.th. like that.

Tests with other openssl versions (Linux):
```
1.0.2g-vanilla: fails the same way
1.0.2e-pm: works
1.1.0-pre6-dev: bails out with an error
```

1.0.2e-pm was the one I uploaded to github last year.

**Comment by seccubus** [comment]
@drwetter basically the test is doing what is should do, detection an incorrect result.

There are three possible ways to adress this:
* Comment out the test as you did (which means no further checking for a unwanted result)
* Adjust the unit test so that the incorrect result is also ok
* Fix the script to use openssl 1.0.2e-pm which works as expected
* Fix the error itself

Each way has it's own drawbacks. I think the option of adjusting the unit test to also agree with the other result is the best for now.

**Comment by drwetter** [comment]
Am 8. August 2016 10:47:16 MESZ, schrieb Frank Breedijk <notifications@github.com>:
>@drwetter basically the test is doing what is should do, detection an
>incorrect result.
>
>There are three possible ways to adress this:
>* Comment out the test as you did (which means no further checking for
>a unwanted result)
>* Adjust the unit test so that the incorrect result is also ok
>* Fix the script to use openssl 1.0.2e-pm which works as expected
>* Fix the error itself
>
>Each way has it's own drawbacks. I think the option of adjusting the
>unit test to also agree with the other result is the best for now.


I vote for 4). But there will be a reality check. ;-) 

@rsalz says it's correct behavior and the certificate is not ok. 

I try to contacts the badssl guys later again.

Thx, Dirk


-- 
Sent from my mobile. Excuse my brevity&typos+the phone's autocorrection

**Comment by drwetter** [comment]
Thx everybody helping here, especially @dcooper16 arguing also with openssl and badssl!

**Comment by DDvO** [comment]
This issue has been fixed today in OpenSSL (see also further discussion on https://github.com/openssl/openssl/issues/1418) and should be available soon.

**Comment by drwetter** [comment]
Thanks @DDvO for the heads up! Last message from @ levitte looks different though.

**Comment by DDvO** [comment]
I've meanwhile proposed a new, more thorouh fix: https://github.com/openssl/openssl/pull/10587

**Comment by drwetter** [comment]
I couldn't avoid noticing the discussion ;-) -- there were lots of mails in my inbox of which I only read a few yet

**Comment by DDvO** [comment]
Finally, a fix for this issue has been merged in both the OpenSSL `master` branch (to be released as version 3.0 in Q4 of 2020)
and `OpenSSL_1_1_1-stable` (to be released with the next letter release, 1.1.1h).

**Comment by drwetter** [comment]
Thanks for the heads up
